src/store/chatStore.ts

```diff
@@ -398,33 +412,143 @@ const sendFile = async (chatId, selectedFile, isBlob = false, isRecording = fals
+    const blob = new Blob([arr], { type: 'image/png' });
+
+    const uuid = uuidv4();
+
+    const CancelToken = axios.CancelToken;
+    const source = CancelToken.source();
 
     try {
-        await axios.post(`${config.baseUrl}api/files/${chatId}/${id}`, formData, {
+        imageUploadQueue.value.push({
+            id: uuid,
+            cancelToken: source,
+            error: false,
+            retry: false,
+            error_message: '',
+            type: selectedFile.type,
+            isImage: selectedFile.type.includes('image'),
+            title: selectedFile.name,
+            data: window.URL.createObjectURL(blob),
+            time: new Date(),
+            loaded: 0,
+            total: selectedFile.total,
+            chatId: chatId,
+            selectedFile: selectedFile
+        });
+
+        await axios.post(`${config.baseUrl}api/files/${chatId}/${uuid}`, formData, {
             headers: {
                 'Content-Type': 'multipart/form-data',
             },
+            cancelToken: source.token,
+            onUploadProgress: ({ loaded: progress, total }) => {
+                imageUploadQueue.value = imageUploadQueue.value.map(item => {
+                    if (item.id === uuid) {
+                        return {
+                            ...item,
+                            loaded: progress,
+                            total: total,
+                        };
+                    }
+                    return item;
+                });
+            },
         });
-        console.log('File uploaded.');
-    } catch (e) {
-        let errorBody = '';
-        if (e.message == 'Request failed with status code 413') {
-            errorBody = 'ERROR: File exceeds 100MB limit!';
-        } else {
-            errorBody = 'ERROR: File failed to send!';
-        }
-
-        const failedMsg = {
-            ...msgToSend,
-            body: errorBody,
-            type: 'STRING',
-        };
-        addMessage(chatId, failedMsg);
+            } catch (e) {
+        sendFileCatch(e,uuid)
     }
 };
 
+const sendFileCatch = (e,uuid) => {
+    let errorBody = '';
+    if (e.message == 'Request failed with status code 413') {
+        //!TODO Upload limit
+        errorBody = 'ERROR: File exceeds 20MB limit!';
+        imageUploadQueue.value = imageUploadQueue.value.map(el => {
+            if (uuid === el.id) {
+                return { ...el, error_message: 'File exceeds 20MB limit!', error: true };
+            }
+            return el;
+        });
+    } else {
+
+        errorBody = 'ERROR: File failed to send!';
+        imageUploadQueue.value = imageUploadQueue.value.map(el => {
+            if (uuid === el.id) {
+                return { ...el, error_message: 'File failed to send', error: true, retry: true };
+            }
+            return el;
+        });
+
+    }
+}
+
+export const retrySendFile = async (file) => {
+    if(!file) return;
```

Comment:
don't check if your input is valid, use typescript to your advantage and type the parameter, and make sure you never can call this function without the right data